### PR TITLE
[BugFix] Fix area for image placeholder in graphql response.

### PR DESCRIPTION
**Pull request by VoronoyAlexandr**
### Description (*)
PR for issue #239 Fix resolving of image placeholder for Product without images.
Changed area to "frontend" instead "graphql" for placeholder asset.

### Fixed Issues (if relevant)

1. magento/graphql-ce#239: Fix resolving of image placeholder for Product without images

### Manual testing scenarios (*)
Request:
```
{
  products (search: "Test") {
    
    items {
        id
        name
      small_image {
        url
      }
    }
  }
}
```
Response:
```
{
  "data": {
    "products": {
      "items": [
        {
          "id": 1201,
          "name": "Test",
          "small_image": {
            "url": "http://graphql-ce.loc/static/version1541944966/frontend/Magento/luma/en_US/Magento_Catalog/images/product/placeholder/small_image.jpg"
          }
        }
      ]
    }
  }
}
```

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Product/ProductImage/Url.php [hunk 1]
```diff
@@ -71,7 +79,15 @@ private function getImageUrl(string $imageType, ?string $imagePath): string
         $image = $this->productImageFactory->create();
         $image->setDestinationSubdir($imageType)
             ->setBaseFile($imagePath);
-        $imageUrl = $image->getUrl();
+
+        $imageUrl = $image->isBaseFilePlaceholder()
```
Comment (paliarush) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Product/ProductImage/Url.php]: @naydav could we get someone to test this solution please?

**Review by paliarush — APPROVED** [review]

**Comment by naydav** [comment]
Also, need to unskip test
\Magento\GraphQl\Catalog\ProductImageTest::testProductWithoutBaseImage
and add a few asserts

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Product/ProductImage/Url.php (hunk 1 above)
Comment (naydav) [reply app/code/Magento/CatalogGraphQl/Model/Resolver/Product/ProductImage/Url.php]: We already have test, need just unskip it and add a few asserts

**Comment by naydav** [comment]
@VoronoyAlexandr 
Pls, need to add additional asserts to unskiped API functional test

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Product/ProductImage/Url.php [hunk 2]
```diff
@@ -23,14 +24,23 @@ class Url implements ResolverInterface
      * @var ImageFactory
      */
     private $productImageFactory;
+    /**
+     * @var PlaceholderProvider
+     */
+    private $_placeholderProvider;
```
Comment (naydav) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Product/ProductImage/Url.php:30]: Don't use prefix `_` for variables

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Product/ProductImage/Url.php [hunk 3]
```diff
@@ -23,14 +24,23 @@ class Url implements ResolverInterface
...
 
     /**
+     * Url constructor.
```
Comment (naydav) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Product/ProductImage/Url.php:33]: Extra comment

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Product/ProductImage/Url.php [hunk 4]
```diff
@@ -55,23 +65,27 @@ public function resolve(
...
      */
     private function getImageUrl(string $imageType, ?string $imagePath): string
     {
         $image = $this->productImageFactory->create();
         $image->setDestinationSubdir($imageType)
             ->setBaseFile($imagePath);
-        $imageUrl = $image->getUrl();
-        return $imageUrl;
+
+        if ($image->isBaseFilePlaceholder()) {
```
Comment (naydav) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Product/ProductImage/Url.php:85]: Need to cover this logic with API-functional test

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Image/Placeholder.php [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+namespace Magento\CatalogGraphQl\Model\Resolver\Products\DataProvider\Image;
+
+use Magento\Catalog\Model\View\Asset\PlaceholderFactory;
+use Magento\Framework\App\Config\ScopeConfigInterface;
+use Magento\Framework\View\Asset\Repository as AssetRepository;
+use Magento\Framework\View\Design\Theme\ThemeProviderInterface;
+use Magento\Store\Model\StoreManagerInterface;
+
+/**
+ * Class Placeholder
```
Comment (naydav) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Image/Placeholder.php:17]: Provide clear dockblocks for all of classes

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Image/Placeholder.php [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+ * @package Magento\CatalogGraphQl\Model\Resolver\Products\DataProvider\Image
+ */
+class Placeholder
```
Comment (naydav) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Image/Placeholder.php:20]: Looks like this class has a lot of responsibility (like helper in Magento 1)

**Review by naydav — CHANGES_REQUESTED** [review]

File: dev/tests/api-functional/testsuite/Magento/GraphQl/Catalog/ProductImageTest.php [hunk 7]
```diff
@@ -51,7 +51,6 @@ public function testProductWithBaseImage()
      */
     public function testProductWithoutBaseImage()
     {
-        $this->markTestIncomplete('https://github.com/magento/graphql-ce/issues/239');
```
Comment (naydav) [inline dev/tests/api-functional/testsuite/Magento/GraphQl/Catalog/ProductImageTest.php:54]: Need to add additional logic for checking of placeholders

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Image/Placeholder.php [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    {
+        $imageAsset = $this->_placeholderFactory->create(['type' => $imageType]);
+
+        // check if placeholder defined in config
+        if ($imageAsset->getFilePath()) {
+            return $imageAsset->getUrl();
+        }
+
+        return $this->_assetRepository->createAsset(
+            "Magento_Catalog::images/product/placeholder/{$imageType}.jpg",
```
Comment (naydav) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Image/Placeholder.php:82]: Need to check behavior if I changed default placeholder on some custom value via configuration

**Comment by naydav** [comment]
Hello @VoronoyAlexandr,
We still need some changes to finish the PR, will you continue progress? 
Thank you

Please let me know if you need any assistance.

**Comment by VitaliyBoyko** [comment]
Closing this one due to inactivity.

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Image/Placeholder.php (hunk 8 above)
Comment (VitaliyBoyko) [reply app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Image/Placeholder.php:82]: It's done above
if ($imageAsset->getFilePath()) {
            return $imageAsset->getUrl();
        }

Comment (VitaliyBoyko) [reply app/code/Magento/CatalogGraphQl/Model/Resolver/Products/DataProvider/Image/Placeholder.php:82]: We need config fixture to check this

**Pull request closed without merging**